swap/cashout.go

```diff
@@ -30,54 +31,104 @@ import (
 // CashChequeBeneficiaryTransactionCost is the expected gas cost of a CashChequeBeneficiary transaction
 const CashChequeBeneficiaryTransactionCost = 50000
 
-// CashoutProcessor holds all relevant fields needed for processing cashouts
-type CashoutProcessor struct {
-	backend    chain.Backend     // ethereum backend to use
-	privateKey *ecdsa.PrivateKey // private key to use
-}
+// CashoutRequestHandlerID is the handlerID used by the CashoutProcessor for CashoutRequests
+const CashoutRequestHandlerID = "CashoutProcessor_CashoutRequest"
 
 // CashoutRequest represents a request for a cashout operation
 type CashoutRequest struct {
 	Cheque      Cheque         // cheque to be cashed
 	Destination common.Address // destination for the payout
 }
 
-// ActiveCashout stores the necessary information for a cashout in progess
-type ActiveCashout struct {
-	Request         CashoutRequest // the request that caused this cashout
-	TransactionHash common.Hash    // the hash of the current transaction for this request
+// CashoutProcessor holds all relevant fields needed for processing cashouts
+type CashoutProcessor struct {
+	backend              chain.Backend     // ethereum backend to use
+	txScheduler          chain.TxScheduler // transaction queue to use
+	cashoutResultHandler CashoutResultHandler
+	cashoutDone          chan *CashoutRequest
+}
+
+// CashoutResultHandler is an interface which accepts CashChequeResults from a CashoutProcessor
+type CashoutResultHandler interface {
+	// Called by the CashoutProcessor when a CashoutRequest was successfully executed
+	// It will be called again if an error is returned
```

Comment:
They don't. The CashoutProcessor calls this when a CashoutRequest was successfully executed (during the `NotifyReceipt` handler). If an error is returned it will be called again, because the `NotifyReceipt` will run again at some point.

Reply:
Seems to be a very vital feature of the queue and deserves (in my opinion) a somewhat more explicit comment here.

I was following the code to actually verify that the txqueue keeps calling `HandleCashoutResult` if `HandleCashoutResult` returns an error, but could not find this logic:

- `processQueue` calls `processRequest` (directly or via `processActveRequest`). (`txqueue.go:634`)
- When the status is `pending`, `processRequest` calls `txq.waitForActiveTransaction` (`txqueue.go:477`)
- `waitForActiveTransaction` blocks until the tx is mined (or txqueue/tx context times out). If tx is mined, it calls `txq.finalizeRequestConfirmed` (`txqueue.go:610`)
- `finalizeRequestConfirmed` calls `txq.notify`, who on it's turn returns the `trigger` function. The `trigger` function is returned by `notify`, allowing the caller of `notify` to tell the queue that an item is waiting.
- `finalizeRequestConfirmed` calles the `trigger` function, returned by `txq.notify`

And now I lost the track... Can you help me by explaining how the call proceeds from here? How and where is this trigger signal received and how/where can I see that indeed, `HandleCashoutResult` is retried if it failed?

Reply:
All notification queues are processed in their respective go routine started by the `SetHandlers` function. If a handler is registered the trigger will wake the `key, err := notifyQueue.Next(txq.ctx, &item, &txq.lock)` call on line 266.